README.md

# Git mirror of cppcrypto

cppcrypto provides optimized implementations of cryptographic primitives.

Includes sample command-line tools:
- 'digest' - for calculating and verifying file checksum(s) using any of the supported hash algorithms (similar to md5sum or RHash).
- 'cryptor' - for file encryption using Serpent-256 algorithm in AEAD mode.

Check out the [cppcrypto web site](https://sourceforge.net/projects/cppcrypto/) for programming documentation.

Features
- Simple self-explanatory programming interface.
- Hash functions: BLAKE, BLAKE2, Echo, Esch, Grøstl, JH, Kupyna, MD5, SHA-1, SHA-2, SHA-3, SHAKE, Skein, SM3, Streebog, Whirlpool.
- Block ciphers: Rijndael (AES), Anubis, Aria, Camellia, CAST-256, Kalyna, Kuznyechik, Mars, Serpent, Simon-128, SM4, Speck-128, Threefish, Twofish.
- Stream ciphers: HC-128, HC-256, Salsa20/20, Salsa20/12, XSalsa20/20, XSalsa20/12, ChaCha20, ChaCha12, XChaCha20, XChaCha12.
- Encryption modes: CBC, CTR.
- AEAD modes: Encrypt-then-MAC, GCM, OCB, ChaCha-Poly1305, Schwaemm.
- Streaming authenticated encryption with associated data (Streaming AEAD).
- MAC functions: HMAC, Poly1305.
- Key derivation functions: PBKDF2, scrypt, Argon2i/Argon2d/Argon2id, HKDF
- Tested compilers: Visual C++ 2017, Visual C++ 2019, Visual C++ 2022, gcc 12.2.1, clang 13.0.0, clang 15.0.7.
- Tested operating systems: Windows, Linux, FreeBSD, OS X, Solaris
- Includes portable implementations and optimized implementations (using SSE/AVX/etc) for modern CPUs.
- The fastest implementation is selected dynamically at runtime depending on CPU features.
- The only publicly-available implementation of AES-NI acceleration for all 25 Rijndael variants (not only for 3 AES variants).
- The only publicly-available performance-optimized implementation of Kupyna hash function (Ukrainian national standard DSTU 7564:2014).

# Author and reference

- Github page of [Keru Kuro](https://github.com/kerukuro)
- Sourceforge page of [Keru Kuro](https://sourceforge.net/u/kerukuro/profile/)
- Homepage of [cppcrypto](https://sourceforge.net/projects/cppcrypto/) @ Sourceforge

/TR 2024-01-13


doc/ChangeLog.txt
2023-03-12 cppcrypto-0.20

New AEAD modes: aead_etm, gcm, ocb, ietf_chacha_poly, schwaemm.
Streaming AEAD mode.
Removed support for Solaris.

2022-09-17 cppcrypto-0.19

Improved performance of CTR mode.

2022-08-11 cppcrypto-0.18

New hash algorithms: Echo, Esch.

2017-10-30 cppcrypto-0.17

Fixed crash on some older CPUs.

2017-08-29 cppcrypto-0.16

New hash algorithm: SHAKE128, SHAKE256.
New key derivation function: Argon2id.
Added 'cryptor' utility demonstrating how to use cppcrypto for file encryption.
API change: blake, blake2s, blake2b, skein256, skein512, skein1024, sha512, sha3, jh, groestl, streebog now accept output length parameter.
Output length specific classes for these algorithms were removed.

2016-03-21 cppcrypto-0.15

Argon2 implementation updated to match version 1.3 of the specification.
(Old version is still supported for compatibility.)

2016-02-13 cppcrypto-0.14

New block cipher algorithms: Simon, Speck.
New MAC algorithm: Poly1305.
New supported platform: Cygwin (experimental).

2015-12-27 cppcrypto-0.13

New stream cipher algorithms: ChaCha, HC-128, HC-256, Salsa20, XChaCha, XSalsa20.

2015-12-21 cppcrypto-0.12

New block cipher algorithm: Threefish.
New key derivation functions: scrypt, Argon2.

2015-12-15 cppcrypto-0.11

New block cipher algorithm: Kuznyechik, Mars, SM4.
New hash algorithm: BLAKE2.
New supported platform: Solaris (using Solaris Studio 12.4 C++ compiler).

2015-12-09 cppcrypto-0.10

New block cipher algorithm: Aria.
New supported platforms: OS X, FreeBSD (using clang 3.4.1+).

2015-12-04 cppcrypto-0.09

New block cipher algorithms: Camellia, Kalyna.

2015-11-26 cppcrypto-0.08

New block cipher algorithms: CAST6, Serpent, Rijndael with block sizes 160 and 224.

2015-11-15 cppcrypto-0.07

New hash algorithms: JH, MD5, SHA-1, SM3, Streebog.

2015-11-11 cppcrypto-0.06

New block cipher algorithm: Twofish.
New hash algorithm: SHA-3.

2015-11-02 cppcrypto-0.05

New block cipher algorithm: Anubis with key sizes 128, 160, 192, 224, 256, 288, 320.
New block cipher algorithm: Rijndael with block size 192 and key sizes 128, 160, 192, 224, 256.
New key derivation function: PBKDF2.
Encryption modes: CTR.

2015-10-10 cppcrypto-0.04

New block cipher algorithm: Rijndael with block sizes 128, 256 and key sizes 128, 160, 192, 224, 256.
Encryption modes: CBC.
MAC functions: HMAC.

2015-09-24 cppcrypto-0.03

New hash algorithm: Kupyna.

2015-09-12 cppcrypto-0.02

New supported platforms: Linux x86_64 and i586 (using gcc 5.1.1).

2015-09-10 cppcrypto-0.01

Initial version.
Supported hash algorithms: BLAKE, Groestl, SHA-2 (incl. SHA512/224 and SHA512/256),
Skein-256, Skein-512, Skein-1024, Whirlpool.
Supported platforms: Windows (32bit and 64bit).
Supported compilers: Visual C++ 2013, Visual C++ 2015.




cppcrypto/aria.h
/*
This code is written by kerukuro for cppcrypto library (http://cppcrypto.sourceforge.net/)
and released into public domain.
*/

#ifndef CPPCRYPTO_ARIA_H
#define CPPCRYPTO_ARIA_H

#include <stdint.h>
#include "block_cipher.h"

namespace cppcrypto
{

	class aria128 : public block_cipher
	{
	public:
		~aria128();

		size_t blocksize() const override { return 128; }
		size_t keysize() const override { return 128; }
		aria128* clone() const override { return new aria128; }
		void clear() override;

		bool init(const unsigned char* key, block_cipher::direction direction) override;
		void encrypt_block(const unsigned char* in, unsigned char* out) override;
		void decrypt_block(const unsigned char* in, unsigned char* out) override;

	private:
		uint32_t rk[13 * 4];
	};

	class aria256 : public block_cipher
	{
	public:
		~aria256();

		size_t blocksize() const override { return 128; }
		size_t keysize() const override { return 256; }
		aria256* clone() const override { return new aria256; }
		void clear() override;

		bool init(const unsigned char* key, block_cipher::direction direction) override;
		void encrypt_block(const unsigned char* in, unsigned char* out) override;
		void decrypt_block(const unsigned char* in, unsigned char* out) override;

	private:
		uint32_t rk[17 * 4];
	};


	class aria192 : public block_cipher
	{
	public:
		~aria192();

		size_t blocksize() const override { return 128; }
		size_t keysize() const override { return 192; }
		aria192* clone() const override { return new aria192; }
		void clear() override;

		bool init(const unsigned char* key, block_cipher::direction direction) override;
		void encrypt_block(const unsigned char* in, unsigned char* out) override;
		void decrypt_block(const unsigned char* in, unsigned char* out) override;

	private:
		uint32_t rk[15 * 4];
	};

}

#endif

